packages/codec/lib/format/types.ts

```diff
@@ -285,13 +313,211 @@ export interface MagicType {
   //introduce such complexity here, especially as this type is basically just
   //for the debugger
   memberTypes?: {
-    [field: string]: Type
+    [field: string]: Type;
   };
   //may have more optional fields defined in the future
 }
 
-export type ReferenceType = ArrayType | MappingType | StructType | StringType | BytesTypeDynamic;
+export type ReferenceType =
+  | ArrayType
+  | MappingType
+  | StructType
+  | StringType
+  | BytesTypeDynamic;
 
 export interface TypesById {
   [id: string]: UserDefinedType;
-};
+}
+
+function isUserDefinedType(anyType: Type): anyType is UserDefinedType {
```

Comment:
Why are their functions in this file?  Everywhere else we keep them separate, and here we *really* should, I think.  This file should contain types only.

Reply:
I am now of the mind that it doesn't make sense to have types-only modules... like if you do

```javascript
const { Format } = require("@truffle/codec");
```

in Javascript, you get nothing!

It only seems logical that helper functions live alongside the types upon with they operate.

Reply:
> If you do `const { Format } = require("@truffle/codec");` in Javascript, you get nothing!

Don't do that then :P

> It only seems logical that helper functions live alongside the types upon which they operate.

So, what, are you going to combine them elsewhere?  Because so far you've kept them separate everywhere else, with a `types.ts` in each folder.  And combining them seems a bit hard to do with say `Values` (where are all the functions that operate on those? They're all over the place), and meanwhile you have other `Types` functions still living elsewhere (e.g. in `format/utils`, which is where `maketype.ts` lives -- will those functions be moved into `types.ts` as well?).  I mean, if you can make it work, go ahead, but I think splitting it is the easier solution.

Reply:
I've been keeping `types.ts` separate, but importing them into the same namespace. The reason for this is often to avoid circular dependencies. As stated in my PR description, my thinking is that `require("<domain>")` provides a superset over `require("<domain>/types")`

Reply:
Re: MakeType... the distinction I am making is that the functions I added here take a `Format.Types.Type` as an argument, whereas MakeType _returns_ Type objects

Reply:
I believe we've reached consensus on this (at least until you have more desire to debate me on it, I'm guessing :)